priv/rest.ex.eex

```diff
@@ -79,60 +79,52 @@ defmodule <%= context.module_name %> do
   defp request(client, method, path, query, headers, input, options, success_status_code) do
     client = %{client | service: "<%= context.signing_name %>"<%= if context.is_global do %>,
                         region:  "<%= context.credential_scope %>"<% end %>}
     <%= if context.endpoint_prefix == "s3-control" do %>account_id = :proplists.get_value("x-amz-account-id", headers)
     host = build_host(account_id, "<%= context.endpoint_prefix %>", client)<% else %>host = build_host("<%= context.endpoint_prefix %>", client)<% end %>
     url = host
     |> build_url(path, client)
-    |> add_query(query)
+    |> add_query(query, client)
 
     additional_headers = [{"Host", host}, {"Content-Type", "<%= context.content_type %>"}]
     headers = AWS.Request.add_headers(additional_headers, headers)
 
-    payload = encode_payload(input)
+    payload = encode!(client, input)
     headers = AWS.Request.sign_v4(client, method, url, headers, payload)
-    perform_request(method, url, payload, headers, options, success_status_code)
+    perform_request(client, method, url, payload, headers, options, success_status_code)
   end
 
-  defp perform_request(method, url, payload, headers, options, nil) do
-    case HTTPoison.request(method, url, payload, headers, options) do
-      {:ok, %HTTPoison.Response{status_code: 200, body: ""} = response} ->
-        {:ok, response}
-
-      {:ok, %HTTPoison.Response{status_code: status_code, body: body} = response}
-      when status_code == 200 or status_code == 202 or status_code == 204 ->
-        {:ok, <%= context.decode %>, response}
+  defp perform_request(client, method, url, payload, headers, options, nil) do
+    case AWS.Client.request(client, method, url, payload, headers, options) do
+      {:ok, %{status_code: status_code, body: body} = response}
+      when status_code in [200, 202, 204] ->
+        body = if(body != "", do: decode!(client, body))
+        {:ok, body, response}
 
-      {:ok, %HTTPoison.Response{body: body}} ->
-        error = <%= context.decode %>
-        {:error, error}
+      {:ok, %{body: body}} ->
+        {:error, decode!(client, body)}
 
-      {:error, %HTTPoison.Error{reason: reason}} ->
-        {:error, %HTTPoison.Error{reason: reason}}
+      error = {:error, _reason} -> error
     end
   end
 
-  defp perform_request(method, url, payload, headers, options, success_status_code) do
-    case HTTPoison.request(method, url, payload, headers, options) do
-      {:ok, %HTTPoison.Response{status_code: ^success_status_code, body: ""} = response} ->
-        {:ok, %{}, response}
-
-      {:ok, %HTTPoison.Response{status_code: ^success_status_code, body: body} = response} ->
-        {:ok, <%= context.decode %>, response}
+  defp perform_request(client, method, url, payload, headers, options, success_status_code) do
+    case AWS.Client.request(client, method, url, payload, headers, options) do
+      {:ok, %{status_code: ^success_status_code, body: body} = response} ->
```

Comment:
I think you can merge this `perform_request` clause with the one above by doing this:

      {:ok, %{status_code: status_code, body: body} = response}
      when is_nil(success_status_code) and status_code in [200, 202, 204]
      when status_code == success_status_code ->

WDYT?

Reply:
Woot! I was trying to find a way to do this. I didn't know we could pass [multiple guards in the same clause](https://hexdocs.pm/elixir/patterns-and-guards.html#multiple-guards-in-the-same-clause). Nice!